{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook for interactive plots of a.a. mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Snakemake input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_predicted_count = snakemake.params.min_predicted_count\n",
    "clade_synonyms = snakemake.params.clade_synonyms\n",
    "heatmap_minimal_domain = snakemake.params.heatmap_minimal_domain\n",
    "orf1ab_to_nsps = snakemake.params.orf1ab_to_nsps\n",
    "clade_cluster = snakemake.params.clade_cluster\n",
    "cluster_founder = snakemake.params.cluster_founder\n",
    "aamut_by_cluster_csv = snakemake.input.aamut_by_cluster\n",
    "clade_founder_nts_csv = snakemake.input.clade_founder_nts\n",
    "cluster_corr_min_count = snakemake.params.cluster_corr_min_count\n",
    "outdir = snakemake.output.outdir"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "import Bio.Seq\n",
    "import os\n",
    "import yaml\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = alt.data_transformers.disable_max_rows()\n",
    "\n",
    "os.makedirs(outdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define function that gives clade labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clade_label(clade):\n",
    "    if clade in clade_synonyms:\n",
    "        return f\"{clade} ({clade_synonyms[clade]})\"\n",
    "    else:\n",
    "        return clade\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mapping clades cluster name to their founder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cluster_founder_map(cluster):\n",
    "    if cluster in list(cluster_founder.keys()):\n",
    "        return cluster_founder[cluster]\n",
    "    else:\n",
    "        print(\"Cluster not defined\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dataframe with clades founder amino acids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clade_founder_nts = pd.read_csv(clade_founder_nts_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# codon translation table\n",
    "codon_table = {\n",
    "    f\"{nt1}{nt2}{nt3}\": str(Bio.Seq.Seq(f\"{nt1}{nt2}{nt3}\").translate())\n",
    "    for nt1 in \"ACGT\" for nt2 in \"ACGT\" for nt3 in \"ACGT\"\n",
    "}\n",
    "\n",
    "# get clade founder amino-acids\n",
    "clade_founder_aas = (\n",
    "    clade_founder_nts\n",
    "    .query(\"gene != 'noncoding'\")\n",
    "    [[\"clade\", \"gene\", \"codon\", \"codon_site\"]]\n",
    "    .drop_duplicates()\n",
    "    .assign(\n",
    "        gene=lambda x: x[\"gene\"].str.split(\";\"),\n",
    "        codon=lambda x: x[\"codon\"].str.split(\";\"),\n",
    "        codon_site=lambda x: x[\"codon_site\"].str.split(\";\"),\n",
    "    )\n",
    "    .explode([\"gene\", \"codon\", \"codon_site\"])\n",
    "    .assign(\n",
    "        aa=lambda x: x[\"codon\"].map(codon_table),\n",
    "        codon_site=lambda x: x[\"codon_site\"].astype(int),\n",
    "    )\n",
    "    .rename(columns={\"codon_site\": \"site\", \"aa\": \"amino acid\"})\n",
    "    .drop(columns=\"codon\")\n",
    "    .query(\"gene != 'ORF1a'\")  # this is just subset of ORF1ab\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now convert ORF1ab numbers to nsp numbers\n",
    "orf1ab_to_nsps_df = pd.concat(\n",
    "    [\n",
    "        pd.DataFrame(\n",
    "            [(i, i - start + 1) for i in range(start, end + 1)],\n",
    "            columns=[\"ORF1ab_site\", \"nsp_site\"],\n",
    "        ).assign(nsp=nsp).drop_duplicates()\n",
    "        for nsp, (start, end) in orf1ab_to_nsps.items()\n",
    "    ],\n",
    "    ignore_index=True,\n",
    ")\n",
    "\n",
    "clade_founder_aas = pd.concat(\n",
    "    [\n",
    "        clade_founder_aas,\n",
    "        (\n",
    "            clade_founder_aas\n",
    "            .query(\"gene == 'ORF1ab'\")\n",
    "            .merge(\n",
    "                orf1ab_to_nsps_df,\n",
    "                left_on=\"site\",\n",
    "                right_on=\"ORF1ab_site\",\n",
    "                validate=\"many_to_one\",\n",
    "            )\n",
    "            .drop(columns=[\"gene\", \"ORF1ab_site\", \"site\"])\n",
    "            .rename(columns={\"nsp_site\": \"site\", \"nsp\": \"gene\"})\n",
    "        ),\n",
    "    ],\n",
    "    ignore_index=False,\n",
    ").drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clade</th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>amino acid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>19A</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>19A</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>2</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>19A</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>3</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>19A</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>4</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>19A</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>5</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213055</th>\n",
       "      <td>23F</td>\n",
       "      <td>nsp16</td>\n",
       "      <td>298</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213056</th>\n",
       "      <td>24A</td>\n",
       "      <td>nsp16</td>\n",
       "      <td>298</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213057</th>\n",
       "      <td>24B</td>\n",
       "      <td>nsp16</td>\n",
       "      <td>298</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213058</th>\n",
       "      <td>24C</td>\n",
       "      <td>nsp16</td>\n",
       "      <td>298</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213059</th>\n",
       "      <td>24E</td>\n",
       "      <td>nsp16</td>\n",
       "      <td>298</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>508110 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       clade    gene  site amino acid\n",
       "265      19A  ORF1ab     1          M\n",
       "268      19A  ORF1ab     2          E\n",
       "271      19A  ORF1ab     3          S\n",
       "274      19A  ORF1ab     4          L\n",
       "277      19A  ORF1ab     5          V\n",
       "...      ...     ...   ...        ...\n",
       "213055   23F   nsp16   298          N\n",
       "213056   24A   nsp16   298          N\n",
       "213057   24B   nsp16   298          N\n",
       "213058   24C   nsp16   298          N\n",
       "213059   24E   nsp16   298          N\n",
       "\n",
       "[508110 rows x 4 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clade_founder_aas.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read-in input dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aamut = pd.read_csv(aamut_by_cluster_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clust_fnd = list(cluster_founder.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clade</th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>amino acid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30168</th>\n",
       "      <td>20A</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30171</th>\n",
       "      <td>20A</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>2</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30174</th>\n",
       "      <td>20A</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>3</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30177</th>\n",
       "      <td>20A</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>4</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30180</th>\n",
       "      <td>20A</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>5</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212982</th>\n",
       "      <td>21K</td>\n",
       "      <td>nsp16</td>\n",
       "      <td>296</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213001</th>\n",
       "      <td>20A</td>\n",
       "      <td>nsp16</td>\n",
       "      <td>297</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213012</th>\n",
       "      <td>21K</td>\n",
       "      <td>nsp16</td>\n",
       "      <td>297</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213031</th>\n",
       "      <td>20A</td>\n",
       "      <td>nsp16</td>\n",
       "      <td>298</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213042</th>\n",
       "      <td>21K</td>\n",
       "      <td>nsp16</td>\n",
       "      <td>298</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33874 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       clade    gene  site amino acid\n",
       "30168    20A  ORF1ab     1          M\n",
       "30171    20A  ORF1ab     2          E\n",
       "30174    20A  ORF1ab     3          S\n",
       "30177    20A  ORF1ab     4          L\n",
       "30180    20A  ORF1ab     5          V\n",
       "...      ...     ...   ...        ...\n",
       "212982   21K   nsp16   296          V\n",
       "213001   20A   nsp16   297          N\n",
       "213012   21K   nsp16   297          N\n",
       "213031   20A   nsp16   298          N\n",
       "213042   21K   nsp16   298          N\n",
       "\n",
       "[33874 rows x 4 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clust_founder_aas = clade_founder_aas.query(\"clade in @clust_fnd\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Adding clade columns to `aamut`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aamut = aamut.assign(clade=lambda x: x['cluster'].map(cluster_founder_map))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Retain only mutations from the cluster founder amino acids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aamut_cl_fnd = (\n",
    "    aamut\n",
    "    .rename(columns={'aa_site':'site', 'clade_founder_aa':'amino acid'})\n",
    "    .merge(clust_founder_aas, on=['clade', 'gene', 'site', 'amino acid'], how='inner', validate='many_to_one')\n",
    "    .rename(columns={'amino acid': 'ref_aa'})\n",
    "    .drop(columns=['clade'])\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert sum(aamut_cl_fnd.groupby(['cluster', 'gene', 'site']).apply(lambda x: len(x.ref_aa.unique()) != 1)) == 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Adding Pango lineage to `clade` column in `clust_founder_aas`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clust_founder_aas = clust_founder_aas.assign(clade=lambda x: x[\"clade\"].map(clade_label))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scatter plot of fitness effects\n",
    "\n",
    "Plot correlation in mutation fitness effect ($\\Delta f_{xy}$ values) among cluster of clades with large numbers of counts. We just plot ORF1ab and not its constituent nsps to avoid double counting. We only plot amino-acid mutations (not synonymous ones)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plotting function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_corr_scatters(corr_df_tidy):\n",
    "    \"Plot set of correlation scatters.\"\"\"\n",
    "    \n",
    "    subsets = corr_df_tidy[\"subset\"].unique()\n",
    "    genes = corr_df_tidy[\"gene\"].unique()\n",
    "    \n",
    "    corr_df_wide = pd.merge(\n",
    "        *[\n",
    "            corr_df_tidy\n",
    "            .pivot_table(\n",
    "                index=[\"gene\", \"aa_mutation\"],\n",
    "                values=prop,\n",
    "                columns=\"subset\",\n",
    "            )\n",
    "            .reset_index()\n",
    "            .rename(columns={subset: f\"{prop} {subset}\" for subset in subsets})\n",
    "            for prop in [\"fitness\", \"predicted_count\"]\n",
    "        ]\n",
    "    )\n",
    "    \n",
    "    fitness_min = corr_df_tidy[\"fitness\"].min()\n",
    "    fitness_max = corr_df_tidy[\"fitness\"].max()\n",
    "\n",
    "    gene_selection = alt.selection_multi(\n",
    "        fields=[\"gene\"], bind=\"legend\",\n",
    "    )\n",
    "\n",
    "    predicted_count_selection = alt.selection_single(\n",
    "        bind=alt.binding_range(\n",
    "            min=1,\n",
    "            max=min(5 * min_predicted_count, corr_df_tidy[\"predicted_count\"].quantile(0.9)),\n",
    "            step=1,\n",
    "            name=\"minimum expected count\",\n",
    "        ),\n",
    "        fields=[\"cutoff\"],\n",
    "        init={\"cutoff\": min_predicted_count},\n",
    "    )\n",
    "    \n",
    "    highlight = alt.selection_single(\n",
    "        on=\"mouseover\",\n",
    "        fields=[\"gene\", \"aa_mutation\"],\n",
    "        empty=\"none\",\n",
    "    )\n",
    "    \n",
    "    corr_charts = []\n",
    "    base_chart = alt.Chart(corr_df_wide)\n",
    "    for subset1, subset2 in itertools.combinations(subsets, 2):\n",
    "        base = (\n",
    "            base_chart\n",
    "            .encode(\n",
    "                x=alt.X(\n",
    "                    f\"fitness {subset1}\",\n",
    "                    title=f\"{subset1} fitness effect\",\n",
    "                    scale=alt.Scale(domain=(fitness_min, fitness_max), nice=False),\n",
    "                ),\n",
    "                y=alt.Y(\n",
    "                    f\"fitness {subset2}\",\n",
    "                    title=f\"{subset2} fitness effect\",\n",
    "                    scale=alt.Scale(domain=(fitness_min, fitness_max), nice=False),\n",
    "                ),\n",
    "                tooltip=[\n",
    "                    \"gene\",\n",
    "                    \"aa_mutation\",\n",
    "                    alt.Tooltip(\n",
    "                        f\"fitness {subset1}\", title=f\"{subset1} fitness effect\",\n",
    "                    ),\n",
    "                    alt.Tooltip(\n",
    "                        f\"fitness {subset2}\", title=f\"{subset2} fitness effect\",\n",
    "                    ),\n",
    "                    f\"predicted_count {subset1}\",\n",
    "                    f\"predicted_count {subset2}\",\n",
    "                ],\n",
    "            )\n",
    "            .mark_circle(opacity=0.3)\n",
    "            .properties(width=200, height=200)\n",
    "            .transform_filter(gene_selection)\n",
    "            .transform_filter(\n",
    "                (alt.datum[f\"predicted_count {subset1}\"] >= predicted_count_selection[\"cutoff\"] - 1e-6)\n",
    "                & (alt.datum[f\"predicted_count {subset2}\"] >= predicted_count_selection[\"cutoff\"] - 1e-6)\n",
    "            )\n",
    "        )\n",
    "    \n",
    "        scatter = (\n",
    "            base\n",
    "            .encode(\n",
    "                color=alt.Color(\n",
    "                    \"gene\",\n",
    "                    scale=alt.Scale(\n",
    "                        domain=genes,\n",
    "                        range=[\"#5778a4\"] * len(genes),\n",
    "                    ),\n",
    "                    legend=alt.Legend(\n",
    "                        symbolOpacity=1,\n",
    "                        orient=\"bottom\",\n",
    "                        title=\"click / shift-click to select specific genes to show\",\n",
    "                        titleLimit=500,\n",
    "                        columns=6,\n",
    "                    ),\n",
    "                ),\n",
    "                size=alt.condition(highlight, alt.value(85), alt.value(30)),\n",
    "                opacity=alt.condition(highlight, alt.value(1), alt.value(0.3)),\n",
    "                strokeWidth=alt.condition(highlight, alt.value(1.5), alt.value(0)),\n",
    "            )\n",
    "            .mark_circle(stroke=\"black\")\n",
    "        )\n",
    "    \n",
    "        # regression line and correlation coefficient: https://stackoverflow.com/a/60239699\n",
    "        line = (\n",
    "            base\n",
    "            .transform_regression(\n",
    "                f\"fitness {subset1}\",\n",
    "                f\"fitness {subset2}\",\n",
    "                extent=(fitness_min, fitness_max),\n",
    "            )\n",
    "            .mark_line(color=\"orange\", clip=True)\n",
    "        )\n",
    "    \n",
    "        params_r = (\n",
    "            base\n",
    "            .transform_regression(\n",
    "                f\"fitness {subset1}\",\n",
    "                f\"fitness {subset2}\",\n",
    "                params=True,\n",
    "            )\n",
    "            .transform_calculate(\n",
    "                r=alt.expr.sqrt(alt.datum[\"rSquared\"]),\n",
    "                label='\"r = \" + format(datum.r, \".3f\")',\n",
    "            )\n",
    "            .mark_text(align=\"left\", color=\"orange\", fontWeight=\"bold\")\n",
    "            .encode(\n",
    "                x=alt.value(5),\n",
    "                y=alt.value(8),\n",
    "                text=alt.Text(\"label:N\"),\n",
    "            )\n",
    "        )\n",
    "        \n",
    "        # show number of points\n",
    "        params_n = (\n",
    "            base\n",
    "            .transform_filter(\n",
    "                (~alt.expr.isNaN(alt.datum[f\"fitness {subset1}\"]))\n",
    "                & (~alt.expr.isNaN(alt.datum[f\"fitness {subset2}\"]))\n",
    "            )\n",
    "            .transform_calculate(dummy=alt.datum[f\"fitness {subset1}\"])\n",
    "            .transform_aggregate(n=\"valid(dummy)\")\n",
    "            .transform_calculate(label='\"n = \" + datum.n')\n",
    "            .mark_text(align=\"left\", color=\"orange\", fontWeight=\"bold\")\n",
    "            .encode(\n",
    "                x=alt.value(5),\n",
    "                y=alt.value(20),\n",
    "                text=alt.Text(\"label:N\"),\n",
    "            )\n",
    "        )\n",
    "    \n",
    "        chart = (\n",
    "            (scatter + line + params_r + params_n)\n",
    "            .add_selection(gene_selection)\n",
    "            .add_selection(predicted_count_selection)\n",
    "            .add_selection(highlight)\n",
    "        )\n",
    "    \n",
    "        corr_charts.append(chart)\n",
    "    \n",
    "    ncols = 4\n",
    "    rows = []\n",
    "    for i in range(0, len(corr_charts), ncols):\n",
    "        rows.append(alt.hconcat(*corr_charts[i: i + ncols]))\n",
    "    corr_chart = alt.vconcat(*rows).configure_axis(grid=False)\n",
    "    return corr_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Scatter for probabilistic fitness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clade_corr_df = (\n",
    "    aamut_cl_fnd\n",
    "    .query(\"not subset_of_ORF1ab\")\n",
    "    .query(\"aa_mutation.str[0] != aa_mutation.str[-1]\")\n",
    "    .query(\"cluster != 'all'\")\n",
    "    [['cluster', 'gene', 'aa_mutation', 'predicted_count', 'delta_fitness']]\n",
    "    .assign(\n",
    "        cluster_counts = lambda x: x.groupby(['cluster'])['predicted_count'].transform('sum'),\n",
    "        clade = lambda x: x['cluster'].map(lambda x: clade_label(cluster_founder_map(x))).str.replace(\".\", \"_\", regex=False),\n",
    "        cluster = lambda x: x['cluster'].str.replace(\".\", \"_\", regex=False),\n",
    "    )\n",
    "    .query(\"cluster_counts >= @cluster_corr_min_count\")\n",
    "    .drop(columns=\"cluster_counts\")\n",
    "    .rename(columns={'cluster': 'subset', 'delta_fitness': 'fitness'})\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_corr_chart = plot_corr_scatters(clade_corr_df)\n",
    "cluster_corr_chart_file = os.path.join(outdir, \"cluster_corr_chart.html\")\n",
    "print(f\"Saving to {cluster_corr_chart_file}\")\n",
    "cluster_corr_chart.save(cluster_corr_chart_file)\n",
    "display(cluster_corr_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Scatter for naive fitness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clade_corr_naive_df = (\n",
    "    aamut_cl_fnd\n",
    "    .query(\"not subset_of_ORF1ab\")\n",
    "    .query(\"aa_mutation.str[0] != aa_mutation.str[-1]\")\n",
    "    .query(\"cluster != 'all'\")\n",
    "    [['cluster', 'gene', 'aa_mutation', 'expected_count', 'predicted_count', 'naive_fitness']]\n",
    "    .assign(\n",
    "        cluster_counts = lambda x: x.groupby(['cluster'])['predicted_count'].transform('sum'),\n",
    "        clade = lambda x: x['cluster'].map(lambda x: clade_label(cluster_founder_map(x))).str.replace(\".\", \"_\", regex=False),\n",
    "        cluster = lambda x: x['cluster'].str.replace(\".\", \"_\", regex=False),\n",
    "    )\n",
    "    .query(\"cluster_counts >= @cluster_corr_min_count\")\n",
    "    .drop(columns={'cluster_counts', 'predicted_count'})\n",
    "    .rename(columns={'cluster': 'subset', 'naive_fitness': 'fitness', 'expected_count': 'predicted_count'})\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_corr_chart_naive = plot_corr_scatters(clade_corr_naive_df)\n",
    "cluster_corr_chart_naive_file = os.path.join(outdir, \"cluster_corr_chart_naive.html\")\n",
    "print(f\"Saving to {cluster_corr_chart_naive_file}\")\n",
    "cluster_corr_chart_naive.save(cluster_corr_chart_naive_file)\n",
    "display(cluster_corr_chart_naive)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Heatmaps of mutational effects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plotting function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_aa_fitness(gene, fitness_df, clade_founder_df):\n",
    "    \"\"\"Plot of amino-acid fitness values.\"\"\"\n",
    "    \n",
    "    # biochemically ordered alphabet\n",
    "    aas = tuple(\"RKHDEQNSTYWFAILMVGPC*\")\n",
    "    assert set(fitness_df[\"amino acid\"]).issubset(aas)\n",
    "    \n",
    "    sites = fitness_df[\"site\"].unique().tolist()\n",
    "    \n",
    "    predicted_count_selection = alt.selection_single(\n",
    "        bind=alt.binding_range(\n",
    "            min=1,\n",
    "            max=min(5 * min_predicted_count, fitness_df[\"predicted_count\"].quantile(0.9)),\n",
    "            step=1,\n",
    "            name=\"minimum predicted count\",\n",
    "        ),\n",
    "        fields=[\"cutoff\"],\n",
    "        init={\"cutoff\": min_predicted_count},\n",
    "    )\n",
    "   \n",
    "    site_zoom_brush = alt.selection_interval(\n",
    "        encodings=[\"x\"],\n",
    "        mark=alt.BrushConfig(\n",
    "            stroke=\"gold\", strokeWidth=1.5, fill=\"yellow\", fillOpacity=0.3,\n",
    "        ),\n",
    "    )\n",
    "        \n",
    "    base = (\n",
    "        alt.Chart(fitness_df)\n",
    "        .encode(x=alt.X(\"site:O\", axis=alt.Axis(labelOverlap=\"parity\")))\n",
    "        .transform_filter(\n",
    "            alt.datum[f\"predicted_count\"] >= predicted_count_selection[\"cutoff\"] - 1e-6\n",
    "        )\n",
    "    )\n",
    "    \n",
    "    heatmap_y = alt.Y(\"amino acid\", sort=aas, scale=alt.Scale(domain=aas))\n",
    "    heatmap_base = (\n",
    "        base\n",
    "        .encode(y=heatmap_y)\n",
    "        .properties(width=alt.Step(12), height=alt.Step(12))\n",
    "    )\n",
    "    \n",
    "    # background fill for missing values in heatmap, imputing dummy stat\n",
    "    # to get all cells\n",
    "    heatmap_bg = (\n",
    "        heatmap_base\n",
    "        .transform_impute(\n",
    "            impute=\"_stat_dummy\",\n",
    "            key=\"amino acid\",\n",
    "            keyvals=aas,\n",
    "            groupby=[\"site\"],\n",
    "            value=None,\n",
    "        )\n",
    "        .mark_rect(color=\"gray\", opacity=0.25)\n",
    "    )\n",
    "\n",
    "    # Select fitness for clades cluster\n",
    "    cluster_selection = alt.selection_single(\n",
    "        fields=[\"cluster\"],\n",
    "        bind=alt.binding_select(\n",
    "            options=fitness_df[\"cluster\"].unique(),\n",
    "            name=\"Cluster of clades\",\n",
    "        ),\n",
    "        init={\"cluster\": fitness_df[\"cluster\"].unique().tolist()[0]},\n",
    "    )\n",
    "\n",
    "    # place X values at \"wildtype\"\n",
    "    wildtype_clade_selection = alt.selection_single(\n",
    "        fields=[\"clade\"],\n",
    "        bind=alt.binding_select(\n",
    "            options=clade_founder_df[\"clade\"].unique(),\n",
    "            name=\"X denotes wildtype in\",\n",
    "        ),\n",
    "        init={\"clade\": clade_label(cluster_founder_map(fitness_df[\"cluster\"].unique().tolist()[0]))},\n",
    "    )\n",
    "    heatmap_wildtype = (\n",
    "        alt.Chart(clade_founder_df.query(\"site in @sites\"))\n",
    "        .encode(\n",
    "            x=alt.X(\"site:O\"),\n",
    "            y=heatmap_y,\n",
    "        )\n",
    "        .mark_text(text=\"x\", color=\"black\")\n",
    "        .add_selection(wildtype_clade_selection)\n",
    "        .transform_filter(wildtype_clade_selection)\n",
    "        .transform_filter(site_zoom_brush)\n",
    "    )\n",
    "    \n",
    "    # heatmap showing non-filtered amino acids\n",
    "    heatmap_aas = (\n",
    "        heatmap_base\n",
    "        .encode(\n",
    "            color=alt.Color(\n",
    "                \"fitness:Q\",\n",
    "                legend=alt.Legend(\n",
    "                    orient=\"bottom\",\n",
    "                    titleOrient=\"left\",\n",
    "                    gradientLength=150,\n",
    "                    gradientStrokeColor=\"black\",\n",
    "                    gradientStrokeWidth=0.5,\n",
    "                ),\n",
    "                scale=alt.Scale(\n",
    "                    zero=True,\n",
    "                    nice=False,\n",
    "                    type=\"linear\",\n",
    "                    domainMid=0,\n",
    "                    domain=alt.DomainUnionWith(heatmap_minimal_domain),\n",
    "                ),\n",
    "            ),\n",
    "            stroke=alt.value(\"black\"),\n",
    "            tooltip=[\n",
    "                alt.Tooltip(c, format=\".3g\")\n",
    "                if fitness_df[c].dtype == float\n",
    "                else c\n",
    "                for c in fitness_df.columns\n",
    "            ],\n",
    "        )\n",
    "        .mark_rect()\n",
    "        .add_selection(cluster_selection)\n",
    "        .transform_filter(cluster_selection)\n",
    "        .transform_filter(site_zoom_brush)\n",
    "    )\n",
    "\n",
    "    heatmap = (\n",
    "        (heatmap_bg + heatmap_aas + heatmap_wildtype)\n",
    "        .add_selection(predicted_count_selection)\n",
    "        .transform_filter(site_zoom_brush)\n",
    "    )\n",
    "    \n",
    "    # make lineplot\n",
    "    site_statistics = [\"mean\", \"max\", \"min\"]\n",
    "    site_stat = alt.selection_single(\n",
    "        bind=alt.binding_radio(\n",
    "            options=site_statistics,\n",
    "            name=\"site fitness statistic\",\n",
    "        ),\n",
    "        fields=[\"site fitness statistic\"],\n",
    "        init={\"site fitness statistic\": site_statistics[0]},\n",
    "    )\n",
    "    \n",
    "    lineplot = (\n",
    "        base\n",
    "        .transform_filter(alt.datum[\"amino acid\"] != \"*\")\n",
    "        .transform_filter(cluster_selection)\n",
    "        .transform_aggregate(\n",
    "            **{stat: f\"{stat}(fitness)\" for stat in site_statistics},\n",
    "            groupby=[\"cluster\", \"site\"],\n",
    "        )\n",
    "        .transform_fold(\n",
    "            site_statistics,\n",
    "            [\"site fitness statistic\", \"site fitness\"],\n",
    "        )\n",
    "        .add_selection(site_stat)\n",
    "        .add_selection(site_zoom_brush)\n",
    "        .transform_filter(site_stat)\n",
    "        .encode(\n",
    "            y=alt.Y(\"site fitness:Q\", axis=alt.Axis(grid=False)),\n",
    "            tooltip=[\n",
    "                \"site\",\n",
    "                alt.Tooltip(\"site fitness:Q\", format=\".3g\"),\n",
    "                \"site fitness statistic:N\",\n",
    "            ],\n",
    "        )\n",
    "        .mark_area(color=\"black\", opacity=0.7)\n",
    "        .properties(\n",
    "            height=75,\n",
    "            width=min(750, 12 * fitness_df[\"site\"].nunique()),\n",
    "            title=alt.TitleParams(\n",
    "                \"use this site plot to zoom into regions on the heat map\",\n",
    "                anchor=\"start\",\n",
    "                fontWeight=\"normal\",\n",
    "                fontSize=11,\n",
    "            ),\n",
    "        )\n",
    "    )\n",
    "    \n",
    "    show_stop = alt.selection_single(\n",
    "        fields=[\"_dummy\"],\n",
    "        bind=alt.binding_radio(\n",
    "            options=[\"yes\", \"no\"],\n",
    "            name=\"show stop in magenta on top site plot\",\n",
    "        ),\n",
    "        init={\"_dummy\": \"no\"},\n",
    "    )\n",
    "    \n",
    "    stopplot = (\n",
    "        base\n",
    "        .transform_filter(cluster_selection)\n",
    "        .add_selection(show_stop)\n",
    "        .transform_filter(alt.datum[\"amino acid\"] == \"*\")\n",
    "        .transform_calculate(_dummy=\"'yes'\")\n",
    "        .transform_filter(show_stop)\n",
    "        .encode(\n",
    "            y=alt.Y(\"fitness\", title=\"site fitness\"),\n",
    "            color=alt.value(\"#CC79A7\"),\n",
    "            tooltip=[\"site\", alt.Tooltip(\"fitness\", format=\".3g\", title=\"stop fitness\")],\n",
    "        )\n",
    "        .mark_line(point=True, strokeWidth=0.5, strokeDash=[2, 2])\n",
    "    )\n",
    "    \n",
    "    return (\n",
    "        (alt.layer(lineplot, stopplot) & heatmap)\n",
    "        .properties(\n",
    "            title=alt.TitleParams(\n",
    "                f\"estimated fitness of amino acids for SARS-CoV-2 {gene} protein\",\n",
    "                fontSize=15,\n",
    "            ),\n",
    "        )\n",
    "        .resolve_scale(color=\"independent\")\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene, fitness_df in (\n",
    "    aamut_cl_fnd\n",
    "    .rename(columns={'delta_fitness': 'fitness', 'mutant_aa': 'amino acid'})\n",
    "    [['cluster', 'gene', 'site', 'amino acid', 'fitness', 'predicted_count']]\n",
    "    .groupby(\"gene\")\n",
    "):\n",
    "    if gene == \"ORF1ab\":\n",
    "        continue  # do not make a plot for ORF1ab, too big\n",
    "    chart = plot_aa_fitness(gene, fitness_df, clust_founder_aas.query(\"gene == @gene\"))\n",
    "    gene_name_for_file = gene.replace(\" \", \"_\").replace(\"(\", \"\").replace(\")\", \"\")\n",
    "    chartfile = os.path.join(outdir, f\"{gene.split()[0]}.html\")\n",
    "    print(f\"\\nSaving chart for {gene} to {chartfile}\")\n",
    "    chart.save(chartfile)\n",
    "    display(chart)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SARS2-refined-fitness",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
