{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook for computing amino acid fitness mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Snakemake input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "orf1ab_to_nsps=snakemake.params.orf_to_nsps\n",
    "gene_overlaps=snakemake.params.gene_ov\n",
    "fitness_pseudocount=snakemake.params.fit_pseudo\n",
    "ntmut_fit=snakemake.input.ntfit_csv\n",
    "output=snakemake.output.aafit_csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding module folder to system paths\n",
    "module_path = os.path.abspath(os.path.join(\"..\"))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from modules import aamutfit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Columns to be exploded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "explode_cols = [\n",
    "    \"gene\",\n",
    "    \"clade_founder_aa\",\n",
    "    \"mutant_aa\",\n",
    "    \"codon_site\",\n",
    "    \"aa_mutation\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "Read data, then:\n",
    "\n",
    "* Exclude mutations in overlapping reading frames specified for exclusion\n",
    "* Explode dataframe to make a separate line for each gene.\n",
    "* Drop ORF1a, the reason being that after we exclude overlapping reading frame sites there aren't any ORF1a sites not also in ORF1ab.\n",
    "* Aggregate all expected and actual counts for the same amino acid change for each clade / amino-acid mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read-in fitness of nucleotide mutations\n",
    "ntmut_fit = pd.read_csv(ntmut_fit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>nt_site</th>\n",
       "      <th>nt_mutation</th>\n",
       "      <th>gene</th>\n",
       "      <th>codon_site</th>\n",
       "      <th>aa_mutation</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>noncoding</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>predicted_count</th>\n",
       "      <th>actual_count</th>\n",
       "      <th>tau_squared</th>\n",
       "      <th>f_max</th>\n",
       "      <th>left_conf_int</th>\n",
       "      <th>right_conf_int</th>\n",
       "      <th>f_mean</th>\n",
       "      <th>f_st_dev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>1</td>\n",
       "      <td>A1C</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>2.76300</td>\n",
       "      <td>6.229330</td>\n",
       "      <td>0</td>\n",
       "      <td>1.133304</td>\n",
       "      <td>-2.573587</td>\n",
       "      <td>1.732517</td>\n",
       "      <td>1.361119</td>\n",
       "      <td>-2.916866</td>\n",
       "      <td>1.615931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>1</td>\n",
       "      <td>A1T</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>5.85249</td>\n",
       "      <td>2.605232</td>\n",
       "      <td>0</td>\n",
       "      <td>1.528482</td>\n",
       "      <td>-1.907730</td>\n",
       "      <td>1.980310</td>\n",
       "      <td>1.364161</td>\n",
       "      <td>-2.404429</td>\n",
       "      <td>1.708155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>1</td>\n",
       "      <td>A1G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>22.54600</td>\n",
       "      <td>33.485375</td>\n",
       "      <td>0</td>\n",
       "      <td>1.264059</td>\n",
       "      <td>-3.651029</td>\n",
       "      <td>1.574816</td>\n",
       "      <td>1.351596</td>\n",
       "      <td>-3.921000</td>\n",
       "      <td>1.502344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>2</td>\n",
       "      <td>T2A</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>4.85726</td>\n",
       "      <td>1.214439</td>\n",
       "      <td>0</td>\n",
       "      <td>1.432641</td>\n",
       "      <td>-1.428756</td>\n",
       "      <td>2.124832</td>\n",
       "      <td>1.303251</td>\n",
       "      <td>-2.010018</td>\n",
       "      <td>1.777772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>2</td>\n",
       "      <td>T2C</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>19.84960</td>\n",
       "      <td>37.529563</td>\n",
       "      <td>0</td>\n",
       "      <td>1.408045</td>\n",
       "      <td>-3.679257</td>\n",
       "      <td>1.630200</td>\n",
       "      <td>1.296212</td>\n",
       "      <td>-3.980191</td>\n",
       "      <td>1.495832</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cluster  nt_site nt_mutation       gene codon_site aa_mutation  synonymous  \\\n",
       "0    BA.2        1         A1C  noncoding  noncoding   noncoding       False   \n",
       "1    BA.2        1         A1T  noncoding  noncoding   noncoding       False   \n",
       "2    BA.2        1         A1G  noncoding  noncoding   noncoding       False   \n",
       "3    BA.2        2         T2A  noncoding  noncoding   noncoding       False   \n",
       "4    BA.2        2         T2C  noncoding  noncoding   noncoding       False   \n",
       "\n",
       "   noncoding  expected_count  predicted_count  actual_count  tau_squared  \\\n",
       "0       True         2.76300         6.229330             0     1.133304   \n",
       "1       True         5.85249         2.605232             0     1.528482   \n",
       "2       True        22.54600        33.485375             0     1.264059   \n",
       "3       True         4.85726         1.214439             0     1.432641   \n",
       "4       True        19.84960        37.529563             0     1.408045   \n",
       "\n",
       "      f_max  left_conf_int  right_conf_int    f_mean  f_st_dev  \n",
       "0 -2.573587       1.732517        1.361119 -2.916866  1.615931  \n",
       "1 -1.907730       1.980310        1.364161 -2.404429  1.708155  \n",
       "2 -3.651029       1.574816        1.351596 -3.921000  1.502344  \n",
       "3 -1.428756       2.124832        1.303251 -2.010018  1.777772  \n",
       "4 -3.679257       1.630200        1.296212 -3.980191  1.495832  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ntmut_fit.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get only coding mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "ntmut_fit_coding = aamutfit.get_coding(ntmut_fit, gene_overlaps, explode_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>nt_site</th>\n",
       "      <th>nt_mutation</th>\n",
       "      <th>gene</th>\n",
       "      <th>codon_site</th>\n",
       "      <th>aa_mutation</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>noncoding</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>predicted_count</th>\n",
       "      <th>...</th>\n",
       "      <th>tau_squared</th>\n",
       "      <th>f_max</th>\n",
       "      <th>left_conf_int</th>\n",
       "      <th>right_conf_int</th>\n",
       "      <th>f_mean</th>\n",
       "      <th>f_st_dev</th>\n",
       "      <th>is_overlapping</th>\n",
       "      <th>overlap_to_exclude</th>\n",
       "      <th>clade_founder_aa</th>\n",
       "      <th>mutant_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>795</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>266</td>\n",
       "      <td>A266G</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>1</td>\n",
       "      <td>M1V</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>22.54600</td>\n",
       "      <td>33.016702</td>\n",
       "      <td>...</td>\n",
       "      <td>0.885954</td>\n",
       "      <td>-3.757701</td>\n",
       "      <td>1.551756</td>\n",
       "      <td>1.291045</td>\n",
       "      <td>-4.027320</td>\n",
       "      <td>1.458397</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>796</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>266</td>\n",
       "      <td>A266C</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>1</td>\n",
       "      <td>M1L</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2.76300</td>\n",
       "      <td>8.187283</td>\n",
       "      <td>...</td>\n",
       "      <td>1.047344</td>\n",
       "      <td>-2.774285</td>\n",
       "      <td>1.699045</td>\n",
       "      <td>1.394591</td>\n",
       "      <td>-3.096819</td>\n",
       "      <td>1.584339</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>266</td>\n",
       "      <td>A266T</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>1</td>\n",
       "      <td>M1L</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5.85249</td>\n",
       "      <td>1.668554</td>\n",
       "      <td>...</td>\n",
       "      <td>1.416492</td>\n",
       "      <td>-0.003785</td>\n",
       "      <td>1.254955</td>\n",
       "      <td>1.175321</td>\n",
       "      <td>-0.058746</td>\n",
       "      <td>1.221781</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>267</td>\n",
       "      <td>T267C</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>1</td>\n",
       "      <td>M1T</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>19.84960</td>\n",
       "      <td>41.542965</td>\n",
       "      <td>...</td>\n",
       "      <td>1.414540</td>\n",
       "      <td>-2.914588</td>\n",
       "      <td>1.296379</td>\n",
       "      <td>1.175352</td>\n",
       "      <td>-3.048273</td>\n",
       "      <td>1.241749</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>267</td>\n",
       "      <td>T267A</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>1</td>\n",
       "      <td>M1K</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>4.85726</td>\n",
       "      <td>4.712380</td>\n",
       "      <td>...</td>\n",
       "      <td>1.432641</td>\n",
       "      <td>-0.792537</td>\n",
       "      <td>1.223349</td>\n",
       "      <td>1.174102</td>\n",
       "      <td>-0.852807</td>\n",
       "      <td>1.206019</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    cluster  nt_site nt_mutation    gene codon_site aa_mutation  synonymous  \\\n",
       "795    BA.2      266       A266G  ORF1ab          1         M1V       False   \n",
       "796    BA.2      266       A266C  ORF1ab          1         M1L       False   \n",
       "797    BA.2      266       A266T  ORF1ab          1         M1L       False   \n",
       "798    BA.2      267       T267C  ORF1ab          1         M1T       False   \n",
       "799    BA.2      267       T267A  ORF1ab          1         M1K       False   \n",
       "\n",
       "     noncoding  expected_count  predicted_count  ...  tau_squared     f_max  \\\n",
       "795      False        22.54600        33.016702  ...     0.885954 -3.757701   \n",
       "796      False         2.76300         8.187283  ...     1.047344 -2.774285   \n",
       "797      False         5.85249         1.668554  ...     1.416492 -0.003785   \n",
       "798      False        19.84960        41.542965  ...     1.414540 -2.914588   \n",
       "799      False         4.85726         4.712380  ...     1.432641 -0.792537   \n",
       "\n",
       "     left_conf_int  right_conf_int    f_mean  f_st_dev  is_overlapping  \\\n",
       "795       1.551756        1.291045 -4.027320  1.458397           False   \n",
       "796       1.699045        1.394591 -3.096819  1.584339           False   \n",
       "797       1.254955        1.175321 -0.058746  1.221781           False   \n",
       "798       1.296379        1.175352 -3.048273  1.241749           False   \n",
       "799       1.223349        1.174102 -0.852807  1.206019           False   \n",
       "\n",
       "     overlap_to_exclude  clade_founder_aa mutant_aa  \n",
       "795               False                 M         V  \n",
       "796               False                 M         L  \n",
       "797               False                 M         L  \n",
       "798               False                 M         T  \n",
       "799               False                 M         K  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ntmut_fit_coding.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aggregate counts for amino acid mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_counts = aamutfit.aggregate_counts(ntmut_fit_coding, explode_cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Adding naive fitness estimates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "aamutfit.naive_fitness(aa_counts, fitness_pseudocount=fitness_pseudocount)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>clade_founder_aa</th>\n",
       "      <th>mutant_aa</th>\n",
       "      <th>aa_site</th>\n",
       "      <th>aa_mutation</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>predicted_count</th>\n",
       "      <th>actual_count</th>\n",
       "      <th>tau_squared</th>\n",
       "      <th>naive_fitness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>22</td>\n",
       "      <td>A22A</td>\n",
       "      <td>27.39653</td>\n",
       "      <td>28.863105</td>\n",
       "      <td>10</td>\n",
       "      <td>4.393839</td>\n",
       "      <td>-0.977127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>32</td>\n",
       "      <td>A32A</td>\n",
       "      <td>194.79774</td>\n",
       "      <td>103.357180</td>\n",
       "      <td>472</td>\n",
       "      <td>3.655382</td>\n",
       "      <td>0.883512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>36</td>\n",
       "      <td>A36A</td>\n",
       "      <td>121.94270</td>\n",
       "      <td>146.546479</td>\n",
       "      <td>17</td>\n",
       "      <td>5.260903</td>\n",
       "      <td>-1.945442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>41</td>\n",
       "      <td>A41A</td>\n",
       "      <td>121.94270</td>\n",
       "      <td>203.566927</td>\n",
       "      <td>76</td>\n",
       "      <td>5.492444</td>\n",
       "      <td>-0.470352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>D</td>\n",
       "      <td>22</td>\n",
       "      <td>A22D</td>\n",
       "      <td>12.63590</td>\n",
       "      <td>12.906202</td>\n",
       "      <td>0</td>\n",
       "      <td>1.567256</td>\n",
       "      <td>-3.268496</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cluster gene clade_founder_aa mutant_aa  aa_site aa_mutation  \\\n",
       "0    BA.2    E                A         A       22        A22A   \n",
       "1    BA.2    E                A         A       32        A32A   \n",
       "2    BA.2    E                A         A       36        A36A   \n",
       "3    BA.2    E                A         A       41        A41A   \n",
       "4    BA.2    E                A         D       22        A22D   \n",
       "\n",
       "   expected_count  predicted_count  actual_count  tau_squared  naive_fitness  \n",
       "0        27.39653        28.863105            10     4.393839      -0.977127  \n",
       "1       194.79774       103.357180           472     3.655382       0.883512  \n",
       "2       121.94270       146.546479            17     5.260903      -1.945442  \n",
       "3       121.94270       203.566927            76     5.492444      -0.470352  \n",
       "4        12.63590        12.906202             0     1.567256      -3.268496  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dataframe with refined fitness estimates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_fit = aamutfit.aa_fitness(ntmut_fit_coding, explode_cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition to the entries for full ORF1ab, we also want to have mutations numbered by the nsp naming.\n",
    "\n",
    "First, make a data frame that converts the numbering:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "orf1ab_to_nsps_df = aamutfit.map_orf1ab_to_nsps(orf1ab_to_nsps)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we add to our dataframes that have ORF1ab the estimates for the nsp proteins. Note that these means mutations in both ORF1ab and nsp show up **twice** in the data frame with different names, so we add a column to indicate which genes are a subset of ORF1ab:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_counts = aamutfit.add_nsps(aa_counts, orf1ab_to_nsps_df)\n",
    "aa_fit = aamutfit.add_nsps(aa_fit, orf1ab_to_nsps_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge counts and fitness dataframes and write to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "aamut_fitness = aamutfit.merge_aa_df(aa_fit, aa_counts, explode_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "      <th>clade_founder_aa</th>\n",
       "      <th>mutant_aa</th>\n",
       "      <th>aa_site</th>\n",
       "      <th>aa_mutation</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>predicted_count</th>\n",
       "      <th>actual_count</th>\n",
       "      <th>tau_squared</th>\n",
       "      <th>naive_fitness</th>\n",
       "      <th>subset_of_ORF1ab</th>\n",
       "      <th>delta_fitness</th>\n",
       "      <th>uncertainty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>22</td>\n",
       "      <td>A22A</td>\n",
       "      <td>27.39653</td>\n",
       "      <td>28.863105</td>\n",
       "      <td>10</td>\n",
       "      <td>4.393839</td>\n",
       "      <td>-0.977127</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.238764</td>\n",
       "      <td>0.750704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>32</td>\n",
       "      <td>A32A</td>\n",
       "      <td>194.79774</td>\n",
       "      <td>103.357180</td>\n",
       "      <td>472</td>\n",
       "      <td>3.655382</td>\n",
       "      <td>0.883512</td>\n",
       "      <td>False</td>\n",
       "      <td>0.518985</td>\n",
       "      <td>0.575201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>36</td>\n",
       "      <td>A36A</td>\n",
       "      <td>121.94270</td>\n",
       "      <td>146.546479</td>\n",
       "      <td>17</td>\n",
       "      <td>5.260903</td>\n",
       "      <td>-1.945442</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.777703</td>\n",
       "      <td>0.654703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>41</td>\n",
       "      <td>A41A</td>\n",
       "      <td>121.94270</td>\n",
       "      <td>203.566927</td>\n",
       "      <td>76</td>\n",
       "      <td>5.492444</td>\n",
       "      <td>-0.470352</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.823123</td>\n",
       "      <td>0.624146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>D</td>\n",
       "      <td>22</td>\n",
       "      <td>A22D</td>\n",
       "      <td>12.63590</td>\n",
       "      <td>12.906202</td>\n",
       "      <td>0</td>\n",
       "      <td>1.567256</td>\n",
       "      <td>-3.268496</td>\n",
       "      <td>False</td>\n",
       "      <td>-3.319033</td>\n",
       "      <td>1.571656</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cluster gene clade_founder_aa mutant_aa  aa_site aa_mutation  \\\n",
       "0    BA.2    E                A         A       22        A22A   \n",
       "1    BA.2    E                A         A       32        A32A   \n",
       "2    BA.2    E                A         A       36        A36A   \n",
       "3    BA.2    E                A         A       41        A41A   \n",
       "4    BA.2    E                A         D       22        A22D   \n",
       "\n",
       "   expected_count  predicted_count  actual_count  tau_squared  naive_fitness  \\\n",
       "0        27.39653        28.863105            10     4.393839      -0.977127   \n",
       "1       194.79774       103.357180           472     3.655382       0.883512   \n",
       "2       121.94270       146.546479            17     5.260903      -1.945442   \n",
       "3       121.94270       203.566927            76     5.492444      -0.470352   \n",
       "4        12.63590        12.906202             0     1.567256      -3.268496   \n",
       "\n",
       "   subset_of_ORF1ab  delta_fitness  uncertainty  \n",
       "0             False      -1.238764     0.750704  \n",
       "1             False       0.518985     0.575201  \n",
       "2             False      -1.777703     0.654703  \n",
       "3             False      -0.823123     0.624146  \n",
       "4             False      -3.319033     1.571656  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aamut_fitness.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aamut_fitness.to_csv(output, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SARS2-refined-fitness",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
