{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotate clade-wise mutation counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inputs from snakemake"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clade_founder = snakemake.input.clade_founder\n",
    "counts = snakemake.input.counts\n",
    "rna_struct = snakemake.input.rna_struct\n",
    "founder_csv = snakemake.output.founder_csv\n",
    "counts_csv = snakemake.output.counts_csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in Python modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clade Founders"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in clade founder sequences adding a column giving a site's sequence context."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download and read in data clade founder data from jbloomlab/SARS2-mut_fitness/results \n",
    "founder_df = pd.read_csv(clade_founder)\n",
    "founder_df.sort_values(['clade', 'site'], inplace=True)\n",
    "\n",
    "# Get founder seqs\n",
    "founder_seq_dict = {}\n",
    "for (clade, data) in founder_df.groupby('clade'):\n",
    "    founder_seq_dict[clade] = ''.join(data['nt'])\n",
    "\n",
    "# For each row, get the site's 3mer motif in the corresponding founder sequence\n",
    "def get_motif(site, clade):\n",
    "    founder_seq = founder_seq_dict[clade]\n",
    "    return founder_seq[site-2:site+1]\n",
    "min_and_max_sites = [founder_df['site'].min(), founder_df['site'].max()]\n",
    "founder_df['motif'] = founder_df.apply(\n",
    "    lambda row: np.nan if row['site'] in min_and_max_sites \\\n",
    "        else get_motif(row['site'], row['clade']),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "# Add columns giving the reference codon and motif\n",
    "founder_df = founder_df.merge(\n",
    "    (\n",
    "        founder_df[founder_df['clade'] == '19A']\n",
    "        .rename(columns={'codon' : 'ref_codon', 'motif' : 'ref_motif'})\n",
    "    )[['site', 'ref_codon', 'ref_motif']], on='site', how='left'\n",
    ")\n",
    "\n",
    "founder_df.rename(columns={'site': 'nt_site'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "founder_df.to_csv(founder_csv, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in and annotate counts data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in dataframe on actual and expected counts, and add columns with metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in data\n",
    "\n",
    "# Download and read counts data from jbloomlab/SARS2-mut_fitness/results \n",
    "counts_df = pd.read_csv(counts)\n",
    "counts_df = counts_df.query(\"subset == 'all'\")\n",
    "\n",
    "# Add metadata\n",
    "counts_df[['wt_nt', 'mut_nt']] = counts_df['nt_mutation'].str.extract(r'(\\w)\\d+(\\w)')\n",
    "counts_df['mut_type'] = counts_df['wt_nt'] + counts_df['mut_nt']\n",
    "\n",
    "def get_mut_class(row):\n",
    "    if row['synonymous']:\n",
    "        return 'synonymous'\n",
    "    elif row['noncoding']:\n",
    "        return 'noncoding'\n",
    "    elif '*' in row['mutant_aa']:\n",
    "        return 'nonsense'\n",
    "    elif row['mutant_aa'] != row['clade_founder_aa']:\n",
    "        return 'nonsynonymous'\n",
    "    else:\n",
    "        raise ValueError(row['mutant_aa'], row['clade_founder_aa'])\n",
    "\n",
    "counts_df['mut_class'] = counts_df.apply(lambda row: get_mut_class(row), axis=1)\n",
    "\n",
    "# Add column indicating if clade is pre-Omicron or Omicron\n",
    "pre_omicron_clades = [\n",
    "    '20A', '20B', '20C', '20E', '20G', '20H', '20I', '20J', '21C','21I', '21J'\n",
    "]\n",
    "counts_df['pre_omicron_or_omicron'] = counts_df['clade'].apply(\n",
    "    lambda x: 'pre_omicron' if x in pre_omicron_clades else 'omicron'\n",
    ")\n",
    "\n",
    "# Add column indicating if a site is before the light switch boundary\n",
    "def light_switch(mut, site, lb1=13467, lb2=21562):\n",
    "    if mut in [\"AT\", \"CG\", \"GC\"]:\n",
    "        pos_bool = True if site < lb2 else False\n",
    "    elif mut == \"CT\":\n",
    "        pos_bool = True if site < lb1 else False\n",
    "    else:\n",
    "        pos_bool = False\n",
    "\n",
    "    return pos_bool\n",
    "\n",
    "counts_df['nt_site_before_boundary'] = counts_df.apply(lambda x: light_switch(x.mut_type, x.nt_site), axis=1)\n",
    "\n",
    "# Add column indicating whether RNA sites from the Lan, 2022, Nature Comm. structure\n",
    "# are predicted to be paired, using code from Hensel, 2023, biorxiv\n",
    "#filename = '../data/lan_2022/41467_2022_28603_MOESM11_ESM.txt'\n",
    "with open(rna_struct) as f:\n",
    "    lines = [line.rstrip().split() for line in f]\n",
    "paired = np.array([[int(x[0]),int(x[4])] for x in lines[1:]])\n",
    "paired_dict = dict(zip(paired[:,0], paired[:,1]))\n",
    "def assign_ss_pred(site):\n",
    "    if site not in paired_dict:\n",
    "        return 'nd'\n",
    "    elif paired_dict[site] == 0:\n",
    "        return 'unpaired'\n",
    "    else:\n",
    "        return 'paired'\n",
    "counts_df['ss_prediction'] = counts_df['nt_site'].apply(lambda x: assign_ss_pred(x))\n",
    "counts_df['unpaired'] = counts_df['ss_prediction'].apply(lambda x: 1 if x == 'unpaired' else 0)\n",
    "\n",
    "# Add columns giving a site's motif relative to the clade founder\n",
    "# and the reference sequence\n",
    "counts_df = counts_df.merge(\n",
    "    founder_df[['nt_site', 'clade', 'motif', 'ref_motif']],\n",
    "    on = ['nt_site', 'clade'], how='left',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign motif to genome edges\n",
    "nt_1 = counts_df.loc[counts_df.nt_site ==1, 'wt_nt'].unique()\n",
    "for n in nt_1:\n",
    "    counts_df.loc[(counts_df.nt_site ==1) & (counts_df.wt_nt == n), 'motif'] = \"A\" + n + \"T\"\n",
    "counts_df.loc[counts_df.nt_site == 29903, 'motif'] = 'AAA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clade</th>\n",
       "      <th>subset</th>\n",
       "      <th>nt_site</th>\n",
       "      <th>nt_mutation</th>\n",
       "      <th>exclude</th>\n",
       "      <th>masked_in_usher</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>actual_count</th>\n",
       "      <th>clade_founder_nt</th>\n",
       "      <th>gene</th>\n",
       "      <th>...</th>\n",
       "      <th>wt_nt</th>\n",
       "      <th>mut_nt</th>\n",
       "      <th>mut_type</th>\n",
       "      <th>mut_class</th>\n",
       "      <th>pre_omicron_or_omicron</th>\n",
       "      <th>nt_site_before_boundary</th>\n",
       "      <th>ss_prediction</th>\n",
       "      <th>unpaired</th>\n",
       "      <th>motif</th>\n",
       "      <th>ref_motif</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20A</td>\n",
       "      <td>all</td>\n",
       "      <td>1</td>\n",
       "      <td>A1C</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.96873</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>AC</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>pre_omicron</td>\n",
       "      <td>False</td>\n",
       "      <td>unpaired</td>\n",
       "      <td>1</td>\n",
       "      <td>AAT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20A</td>\n",
       "      <td>all</td>\n",
       "      <td>1</td>\n",
       "      <td>A1G</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3.60910</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AG</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>pre_omicron</td>\n",
       "      <td>False</td>\n",
       "      <td>unpaired</td>\n",
       "      <td>1</td>\n",
       "      <td>AAT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20A</td>\n",
       "      <td>all</td>\n",
       "      <td>1</td>\n",
       "      <td>A1T</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.27820</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>AT</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>pre_omicron</td>\n",
       "      <td>True</td>\n",
       "      <td>unpaired</td>\n",
       "      <td>1</td>\n",
       "      <td>AAT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20A</td>\n",
       "      <td>all</td>\n",
       "      <td>2</td>\n",
       "      <td>T2A</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.90342</td>\n",
       "      <td>0</td>\n",
       "      <td>T</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>TA</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>pre_omicron</td>\n",
       "      <td>False</td>\n",
       "      <td>unpaired</td>\n",
       "      <td>1</td>\n",
       "      <td>ATT</td>\n",
       "      <td>ATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20A</td>\n",
       "      <td>all</td>\n",
       "      <td>2</td>\n",
       "      <td>T2C</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3.58890</td>\n",
       "      <td>0</td>\n",
       "      <td>T</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>TC</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>pre_omicron</td>\n",
       "      <td>False</td>\n",
       "      <td>unpaired</td>\n",
       "      <td>1</td>\n",
       "      <td>ATT</td>\n",
       "      <td>ATT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  clade subset  nt_site nt_mutation  exclude  masked_in_usher  expected_count  \\\n",
       "0   20A    all        1         A1C     True             True         0.96873   \n",
       "1   20A    all        1         A1G     True             True         3.60910   \n",
       "2   20A    all        1         A1T     True             True         1.27820   \n",
       "3   20A    all        2         T2A     True             True         0.90342   \n",
       "4   20A    all        2         T2C     True             True         3.58890   \n",
       "\n",
       "   actual_count clade_founder_nt       gene  ... wt_nt mut_nt mut_type  \\\n",
       "0             0                A  noncoding  ...     A      C       AC   \n",
       "1             0                A  noncoding  ...     A      G       AG   \n",
       "2             0                A  noncoding  ...     A      T       AT   \n",
       "3             0                T  noncoding  ...     T      A       TA   \n",
       "4             0                T  noncoding  ...     T      C       TC   \n",
       "\n",
       "   mut_class pre_omicron_or_omicron  nt_site_before_boundary  ss_prediction  \\\n",
       "0  noncoding            pre_omicron                    False       unpaired   \n",
       "1  noncoding            pre_omicron                    False       unpaired   \n",
       "2  noncoding            pre_omicron                     True       unpaired   \n",
       "3  noncoding            pre_omicron                    False       unpaired   \n",
       "4  noncoding            pre_omicron                    False       unpaired   \n",
       "\n",
       "  unpaired motif  ref_motif  \n",
       "0        1   AAT        NaN  \n",
       "1        1   AAT        NaN  \n",
       "2        1   AAT        NaN  \n",
       "3        1   ATT        ATT  \n",
       "4        1   ATT        ATT  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to file\n",
    "counts_df.drop(columns=['subset'], inplace=True)\n",
    "if not os.path.isfile(counts_csv):\n",
    "    counts_df.to_csv(counts_csv, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SARS2-refined-fitness",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
